### ISSUE TITLE
Mass downloading may result in overwriting files and "RateLimitExceeded" errors

### ISSUE THREAD

**Issue by FatalBulletHit** [SteamRE/DepotDownloader#148]
Downloading each manifest of a depot will result in files with the same name (but different versions) being overwritten if `-dir` is not defined for each manifest. However, defining `-dir` for each manifest will result in files already downloaded from other manifests to be downloaded again.

I suggest to define `-dir` by default with `.\downloads\<app id>\<depot id>\<manifest id>` and use the SHA-1 hashes from the manifests to skip a file if it has already been downloaded from another manifest.

I wrote a PowerShell script to achieve this, but I need to download each manifest with the `-manifest-only` switch first and making hundreds of calls will cause a `RateLimitExceeded` error eventually.
I'm not sure if it is possible to mitigate this problem by making one call for multiple depots and/or manifests, but I suggest to make the `-depot` and `-manifest` parameters accept arrays (#147) and implement a new switch to download all manifests of a depot (e.g. `-all-manifests`) which should ideally work with the `-manifest-only` parameter as well.

If the above should be implemented in one way or another, I also suggest to implement a wait time after an `RateLimitExceeded` error.

Here's the script, maybe it helps: https://pastebin.com/4RUqyWrG

**Comment by xPaw** [comment]
Please don't script againt steamdb. depot downloader is already able to tell you the depot name and fetch the latest manifests.

**Comment by FatalBulletHit** [comment]
The script only sends 2 requests to steamdb per depot, may I ask what's wrong with that?
And how would I retrieve all manifests (not just the latest) with DepotDownloader (or anything but steamdb)?

**Comment by xPaw** [comment]
> The script only sends 2 requests to steamdb per depot

To say the least, you could have grabbed name and manifests in a single request.

> And how would I retrieve all manifests (not just the latest)

You can grab the list from steamdb once, and then simply use depotdownloader (or steamcmd) to download newer versions, as you have seeded it by then. There's zero reason to keep hitting steamdb for that.

**Comment by FatalBulletHit** [comment]
> To say the least, you could have grabbed name and manifests in a single request.

True, fixed that.
> You can grab the list from steamdb once, and then simply use depotdownloader (or steamcmd) to download newer versions, as you have seeded it by then. There's zero reason to keep hitting steamdb every 10 minutes for that.

For clarification, the script is meant to be run once and not constantly.
I made it mainly for myself so I could download a bunch of old audio related files as they used to be packed differently (which allowed for better sorting).
It's nothing more but a working concept which I felt like sharing to hopefully improve this great tool.
> As for your other request about multiple depots in single command: #147

I know about that, but this request is mainly about the `-manifest` parameter rather than the `-depot` parameter, I just thought I'd add it for the sake of completeness.

**Comment by azuisleet** [comment]
I don't believe this use case is feasible based on how we process downloads (-all-manifests or multiple manifests for one depot). It definitely wouldn't work in the scenario of specifying a -dir, because all the files from different versions of a depot would overlap.

The default path for depots is indeed strange but reasonable when you explicitly specify a manifest, since the path has the *current* build number instead of whatever it would have been at the time. We simply don't know what the historical build numbers or manifests are.

I have implemented #147, but I have not made any decisions on specifying specific manifests for multiple depots (it is currently only valid to specify one depot and one manifest).

**Comment by azuisleet** [comment]
The 2.4.0 pre-release includes support for specifying multiple depots, but you cannot specify multiple versions of the same depot with different manifests.

If you have any ideas for how this feature might be feasible feel free to re-open this.

**Comment by FatalBulletHit** [comment]
So, I hope I understand this correctly, but the way I handled multiple depots with different versions was to create a subfolder for each depot and in those another subfolder for each manifest (with datetime in the name to allow for chronological sorting).

To overcome the problem of having an array of depot ids and an array of manifest ids at the same time, we could either:
- retrieve a list of all manifests per depot and check which depot contains which manifest ids (that's what I did by retrieving the manifest list from steamdb and would also easily allow for an `-all-manifests` switch).
- brute force it by requesting for each depot every manifest id (don't know if this could cause some spam error)

Both cases assume that all manifest ids are unique, regardless of their depot id (I'm not sure if that's the case).

Last but not least, I created a list of all the file hashes of the files downloaded, to prevent downloading the same files potentially over and over again (I don't know how common this is, but when downloading all the manifests for all the language depots for Rainbow Six Siege this decreased the download size by 80% or so).

Hope that's what you meant, feel free to use my PowerShell script from above as inspiration if it's helpful at all.
